### PR TITLE
(feat) O3-1942: Lift form privileges to useForms hook

### PR DESCRIPTION
**Pull request by Jexsie**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary

 - Each encounter type has associated user privileges (viewPrivilege and editPrivilege. Previously, we used these to filter 
    forms in the [forms component](https://github.com/openmrs/openmrs-esm-patient-chart/blob/40a0b56f6bf6905061a24a93225e6353c0a9d874/packages/esm-patient-forms-app/src/forms/forms.component.tsx#L40); 
   Lifted this logic to be handled directly from the **useForms** hook with the offline awareness logic
## Related Issue

https://issues.openmrs.org/browse/O3-1942

### CODE REVIEW

**Comment by Jexsie** [comment]
cc @ibacher

File: packages/esm-patient-forms-app/src/hooks/use-forms.ts [hunk 1]
```diff
@@ -55,9 +64,15 @@ export function useForms(patientUuid: string, startDate?: Date, endDate?: Date,
   // If this ever becomes a problem for online mode (i.e. if an error should be rendered there when past encounters
   // for determining filled out forms can't be loaded) this should ideally be conditionally controlled via a flag
   // such that the current offline behavior doesn't change.
+  let formsToDisplay = cachedOfflineFormsOnly
+    ? data?.filter((formInfo) => isValidOfflineFormEncounter(formInfo.form, htmlFormEntryForms))
+    : data;
+  formsToDisplay = formsToDisplay?.filter((formInfo) =>
```
Comment (ibacher) [inline packages/esm-patient-forms-app/src/hooks/use-forms.ts:70]:
```suggestion
  if (session?.user) {
    formsToDisplay = formsToDisplay?.filter((formInfo) =>
```

File: packages/esm-patient-forms-app/src/forms/forms-list.test.tsx [hunk 2]
```diff
@@ -36,6 +45,10 @@ it('renders a list of forms fetched from the server', async () => {
     showConfigurableForms: false,
   }));
 
+  mockedUserHasAccess.mockImplementation((privilege) => {
```
Comment (ibacher) [inline packages/esm-patient-forms-app/src/forms/forms-list.test.tsx:48]: So, we should have some sort of `expect` in the body of this test that asserts that:

1. At least one of the matching forms is rendered
2. At least one of the non-matching forms is not rendered.

**Review by ibacher — COMMENTED** [review]
@Jexsie Thanks! Looking good, just a couple of small things....

File: packages/esm-patient-forms-app/src/forms/forms-list.test.tsx (hunk 2 above)
Comment (Jexsie) [reply packages/esm-patient-forms-app/src/forms/forms-list.test.tsx:48]: I have added `expect` blocks, but, neglecting the recent changes I have made, all the mocked forms seem not to be rendered in the table. Only three forms were rendered.
Screenshot from 2023-03-21 18-03-52

After checking, the `published: false` seemed to be causing the issue. On changing to true for those forms that were not displaying, only one of the forms changed was displayed. I hope this helps in any way.

File: packages/esm-patient-forms-app/src/hooks/use-forms.ts [hunk 3]
```diff
@@ -67,9 +67,12 @@ export function useForms(patientUuid: string, startDate?: Date, endDate?: Date,
   let formsToDisplay = cachedOfflineFormsOnly
     ? data?.filter((formInfo) => isValidOfflineFormEncounter(formInfo.form, htmlFormEntryForms))
     : data;
-  formsToDisplay = formsToDisplay?.filter((formInfo) =>
-    userHasAccess(formInfo?.form?.encounterType?.editPrivilege?.display, session?.user),
-  );
+
+  if (session?.user) {
+    formsToDisplay = formsToDisplay?.filter((formInfo) =>
+      userHasAccess(formInfo?.form?.encounterType?.editPrivilege?.display, session?.user),
```
Comment (ibacher) [inline packages/esm-patient-forms-app/src/hooks/use-forms.ts:73]: One very small simplification:

```ts
userHasAccess(formInfo?.form?.encounterType?.editPrivilege?.display, session.user)
```

File: packages/esm-patient-forms-app/src/forms/forms-list.test.tsx [hunk 4]
```diff
@@ -57,6 +57,8 @@ it('renders a list of forms fetched from the server', async () => {
   expect(searchbox).toBeInTheDocument();
   expect(screen.getByRole('button', { name: /clear search input/i })).toBeInTheDocument();
   expect(screen.getByRole('table')).toBeInTheDocument();
+  expect(screen.getByRole('cell', { name: /Laboratory Tests/i })).toBeInTheDocument();
+  expect(screen.queryByText('PiusRocks')).toBe(null);
```
Comment (ibacher) [inline packages/esm-patient-forms-app/src/forms/forms-list.test.tsx:61]: `.not.toBeInTheDocument()` is a little more on point, I think?

File: packages/esm-patient-forms-app/src/hooks/use-forms.ts (hunk 3 above)
Comment (Jexsie) [reply packages/esm-patient-forms-app/src/hooks/use-forms.ts:73]: Well @ibacher, the comma is auto-appended on save

Comment (ibacher) [reply packages/esm-patient-forms-app/src/hooks/use-forms.ts:73]: The point was removing the last `?`, not the comma...

**Review by ibacher — APPROVED** [review]

**Pull request merged**